core/bin/dust_api.rs

```diff
@@ -1871,6 +1871,140 @@ async fn databases_tables_list(
     }
 }
 
+#[derive(serde::Deserialize)]
+struct DatabasesRowsUpsertPayload {
+    contents: HashMap<String, serde_json::Value>,
+    truncate: Option<bool>,
+}
+
+async fn databases_rows_upsert(
+    extract::Path((project_id, data_source_id, database_id, table_id)): extract::Path<(
+        i64,
+        String,
+        String,
+        String,
+    )>,
+    extract::Json(payload): extract::Json<DatabasesRowsUpsertPayload>,
+    extract::Extension(state): extract::Extension<Arc<APIState>>,
+) -> (StatusCode, Json<APIResponse>) {
+    let truncate = match payload.truncate {
+        Some(v) => v,
+        None => false,
+    };
+    let project = project::Project::new_from_id(project_id);
+
+    match state
+        .store
+        .batch_upsert_database_rows(
+            &project,
+            &data_source_id,
+            &database_id,
+            &table_id,
+            &payload.contents,
+            truncate,
+        )
+        .await
+    {
+        Err(e) => error_response(
+            StatusCode::INTERNAL_SERVER_ERROR,
+            "internal_server_error",
+            "Failed to upsert database rows",
+            Some(e),
+        ),
+        Ok(()) => (
+            StatusCode::OK,
+            Json(APIResponse {
+                error: None,
+                response: Some(json!({
+                    "success": true
+                })),
+            }),
+        ),
+    }
+}
+
+async fn databases_rows_retrieve(
+    extract::Path((project_id, data_source_id, database_id, table_id, row_id)): extract::Path<(
+        i64,
+        String,
+        String,
+        String,
+        String,
+    )>,
+
+    extract::Extension(state): extract::Extension<Arc<APIState>>,
+) -> (StatusCode, Json<APIResponse>) {
+    let project = project::Project::new_from_id(project_id);
+
+    match state
+        .store
+        .load_database_row(&project, &data_source_id, &database_id, &table_id, &row_id)
+        .await
+    {
+        Err(e) => error_response(
+            StatusCode::INTERNAL_SERVER_ERROR,
+            "internal_server_error",
+            "Failed to upsert database rows",
+            Some(e),
+        ),
+        Ok(row) => (
+            StatusCode::OK,
+            Json(APIResponse {
+                error: None,
+                response: Some(json!({
+                    "row": row
+                })),
+            }),
+        ),
+    }
+}
+
+#[derive(serde::Deserialize)]
+struct DatabasesRowsListQuery {
+    offset: usize,
+    limit: usize,
+    table_id: Option<String>,
```

Comment:
I think it's a bad design to fetch all at once. We never want to do something with 2 rows from different tables and can always parallelize

Reply:
> We never want to do something with 2 rows

We kind of do when we generate the schema from scratch though, no ?